zilliqa/src/consensus.rs

```diff
@@ -1694,22 +1733,43 @@ impl Consensus {
         Ok(())
     }
 
-    pub fn leader(&self, committee: &Committee, view: u64) -> Validator {
+    pub fn leader_at_block(&mut self, block: &Block, view: u64) -> Validator {
+        let curr_root_hash = self.state.root_hash().unwrap();
+        let block_root_hash = block.state_root_hash();
+
+        self.state.set_to_root(H256(block_root_hash.0));
+        let leader = self.leader(view);
+        self.state.set_to_root(H256(curr_root_hash.0));
```

Comment:
Similarly, here also using `state.at_root()` would allow eliminating the `mut`, which would avoid `mut`s up the chain - including changes to the RPCs to require `mut Node` that this PR makes due to this function taking `mut self`